server/src/main/java/org/elasticsearch/index/mapper/JsonFieldParser.java

```diff
@@ -31,53 +31,117 @@
         while (true) {
-            if (openObjects == 0) {
-                return fields;
+            XContentParser.Token token = parser.nextToken();
+            if (token == XContentParser.Token.END_OBJECT) {
+                return;
             }
 
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                assert currentName != null;
+                path.add(currentName);
+                parseObject(parser, path, fields);
+                path.remove();
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                assert currentName != null;
+                parseArray(parser, path, currentName, fields);
+            }  else if (token.isValue()) {
+                String value = parser.text();
+                addField(path, currentName, value, fields);
+            } else if (token == XContentParser.Token.VALUE_NULL) {
+                String value = fieldType.nullValueAsString();
+                if (value != null) {
+                    addField(path, currentName, value, fields);
+                }
+            }
+        }
+    }
+
+    private void parseArray(XContentParser parser,
+                             ContentPath path,
+                             String currentName,
+                             List<IndexableField> fields) throws IOException {
+        while (true) {
             XContentParser.Token token = parser.nextToken();
-            assert token != null;
+            if (token == XContentParser.Token.END_ARRAY) {
+                return;
+            }
 
             if (token == XContentParser.Token.START_OBJECT) {
-                openObjects++;
-            } else if (token == XContentParser.Token.END_OBJECT) {
-                openObjects--;
+                path.add(currentName);
+                parseObject(parser, path, fields);
+                path.remove();
             } else if (token.isValue()) {
                 String value = parser.text();
-                addField(value, fields);
+                addField(path, currentName, value, fields);
             } else if (token == XContentParser.Token.VALUE_NULL) {
                 String value = fieldType.nullValueAsString();
                 if (value != null) {
-                    addField(value, fields);
+                    addField(path, currentName, value, fields);
                 }
             }
         }
     }
 
-    private void addField(String value, List<IndexableField> fields) {
-        if (value.length() <= ignoreAbove) {
-            fields.add(new Field(fieldType.name(), new BytesRef(value), fieldType));
+    private void addField(ContentPath path,
+                          String currentName,
+                          String value,
+                          List<IndexableField> fields) {
+        if (value.length() > ignoreAbove) {
```

Comment:
For prefixed values, the alternative option here would be to check the whole length of the prefixed token, as opposed to just the value. I think that this behavior is more intuitive (and I also don't think we're as concerned about field keys being really long, as opposed to values?)

Reply:
Makes sense to me, I'll make a note on the meta-issue to add a limit.